python/ray/train/torch.py

```diff
@@ -74,6 +86,19 @@ def prepare_model(
         if move_to_device:
             logger.info(f"Moving model to device: {device}")
             model = model.to(device)
+
+        def wrap_forward(forward):
+            @functools.wraps(forward)
+            def wrapper(*args, **kwargs):
+                with autocast(enabled=self.amp):
+                    outputs = forward(*args, **kwargs)
+                assert isinstance(outputs, torch.Tensor)
+                return outputs.float()
+
+            return wrapper
+
+        model.forward = wrap_forward(model.forward)
```

Comment:
Should we not even do wrapping at all if `self.amp` is False?

Reply:
This is purely an implementation detail, right? If `self.amp` is `False`, autocasting is disabled. On line 93 of `torch.py`:  `with autocast(enabled=self.amp)`.

Reply:
It just adds more risk though, right?

If we don't want amp at all
1. Do we still want to do `outputs.float()`?
2. We're relying on `autocast(enabled=False)` to be fully benign

I don't think there's a reason to modify the forward method at all if we don't want to use amp. And in the future, if we do, then we can always revisit this.

Reply:
>  2. Would it be possible to just add a simple example for this? We can just copy the existing linear example call it `torch_linear_amp.py`.

I can add an example. Is the example intended for user consumption? If so, I don't think we should use the linear example, since training will be slower with AMP on a linear dataset.

Also, rather than introducing a new example, I could just add AMP as a config to an existing example. What do you think?

In any case, is it alright if I address this in a follow-up PR?

Reply:
Fixed

Reply:
Fixed